graylog2-server/src/main/java/org/graylog/security/entities/EntityDependencyResolver.java

```diff
@@ -90,27 +76,50 @@ public ImmutableSet<EntityDescriptor> resolve(GRN entity) {
                 //  but in this context we want to return the actual dependent Stream to add additional permissions to.
-                .map(descriptor -> ModelTypes.STREAM_REF_V1.equals(descriptor.type())
-                        ? org.graylog2.contentpacks.model.entities.EntityDescriptor.create(descriptor.id(), ModelTypes.STREAM_V1)
-                        : descriptor)
-                .map(descriptor -> grnRegistry.newGRN(descriptor.type().name(), descriptor.id().id()))
+                .map(cpDescriptor -> ModelTypes.STREAM_REF_V1.equals(cpDescriptor.type())
+                        ? org.graylog2.contentpacks.model.entities.EntityDescriptor.create(cpDescriptor.id(), ModelTypes.STREAM_V1)
+                        : cpDescriptor)
+                .map(cpDescriptor -> grnRegistry.newGRN(cpDescriptor.type().name(), cpDescriptor.id().id()))
                 .filter(dependency -> !entity.equals(dependency)) // Don't include the given entity in dependencies
                 .collect(ImmutableSet.toImmutableSet());
 
-        final Map<GRN, Set<GRN>> targetOwners = grantService.getOwnersForTargets(dependencies);
-
         return dependencies.stream()
-                .map(dependency -> {
-                    String title = entityExcerpts.get(dependency) != null ? entityExcerpts.get(dependency).orElse("unnamed dependency: <" + dependency + ">") : "unknown dependency: <" + dependency + ">";
-                    return EntityDescriptor.create(
-                            dependency,
-                            title,
-                            getOwners(targetOwners.get(dependency))
-                    );
-                })
+                .flatMap(dependency -> descriptorFromGRN(dependency, entityExcerpts()).stream())
                 .collect(ImmutableSet.toImmutableSet());
     }
 
+    private ImmutableMap<GRN, Optional<String>> entityExcerpts() {
+        // TODO: Replace entity excerpt usage with GRNDescriptors once we implemented GRN descriptors for every entity
+        return contentPackService.listAllEntityExcerpts().stream()
+                // TODO: Use the GRNRegistry instead of manually building a GRN. Requires all entity types to be in the registry.
+                .collect(ImmutableMap.toImmutableMap(e -> GRNType.create(e.type().name(), e.type().name() + ":").newGRNBuilder().entity(e.id().id()).build(),
+                        v -> Optional.ofNullable(v.title())));
+    }
+
+    public Optional<EntityDescriptor> descriptorFromGRN(GRN grn) {
+        return descriptorFromGRN(grn, entityExcerpts()).stream().findFirst();
+    }
+
+    private Set<EntityDescriptor> descriptorFromGRN(GRN entity, ImmutableMap<GRN, Optional<String>> entityExcerpts) {
+        return cpDescriptorfromGRN(entity).stream()
+                .map(descriptor -> {
+                    final Map<GRN, Set<GRN>> targetOwners = grantService.getOwnersForTargets(ImmutableSet.of(entity));
```

Comment:
this is now called for every entity. Ideally there could be a convenience method where the targetOwners can be passed in so the owners would have to be resolved only once for all dependencies from above.